src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -441,22 +470,77 @@ private void SetProxyOptions(SafeCurlHandle requestHandle, Uri requestUri)
             }
         }
 
-        private NetworkCredential GetCredentials(ICredentials proxyCredentials, Uri requestUri)
+        private void SetRequestHandleCredentialsOptions(SafeCurlHandle requestHandle, RequestCompletionSource state)
+        {
+            NetworkCredential credentials = GetNetworkCredentials(state.Handler._serverCredentials, state.RequestMessage.RequestUri);
+            if (credentials != null)
+            {
+                if (string.IsNullOrEmpty(credentials.UserName))
+                {
+                    throw new ArgumentException(SR.net_http_argument_empty_string, "UserName");
+                }
+                string userName = string.IsNullOrEmpty(credentials.Domain) ?
+                    credentials.UserName :
+                    string.Format("{0}\\{1}", credentials.Domain, credentials.UserName);
+
+                SetCurlOption(requestHandle, CURLoption.CURLOPT_USERNAME, userName);
+                SetCurlOption(requestHandle, CURLoption.CURLOPT_HTTPAUTH, CURLAUTH.AuthAny);
+                if (credentials.Password != null)
+                {
+                    SetCurlOption(requestHandle, CURLoption.CURLOPT_PASSWORD, credentials.Password);
+                }
+
+                state.NetworkCredential = credentials;
+            }
+        }
+
+        private NetworkCredential GetNetworkCredentials(ICredentials credentials, Uri requestUri)
         {
-            if (proxyCredentials == null)
+            if (_preAuthenticate)
+            {
+                NetworkCredential nc = null;
+                lock (_multiHandle)
+                {
+                    nc = GetCredentials(_credentialCache, requestUri);
+                }
+                if (nc != null)
+                {
+                    return nc;
+                }
+            }
+
+            return GetCredentials(credentials, requestUri);
```

Comment:
Why don't we need locking here?  Should the locking be moved to inside of GetCredentials, so that the locking is encapsulated inside both the AddCredentialToCache and GetCredentials methods?